Show hide taskpane (#1623)
OfficeDev/office-js-docs-pr: docs/excel/show-hide-add-in.md

---
title: Show or hide an Office Add-in in a shared runtime
description: 'Learn how to programmatically hide or show the UI of an add-in while it runs continuously'
ms.date: 03/02/2020
localization_priority: Normal
---

# Show or hide an Office Add-in in a shared runtime (preview)

> Office Add-ins can include any of the following parts:

Reviewer: What about content add-ins?

Author: Content add-ins are not yet supported in shared runtime.

> Office Add-ins can include any of the following parts:

Reviewer 2: Something about the plural sounds weird to me. Consider the following:
```suggestion
An Office Add-in can include any of the following parts:
```

- a task pane
- a UI-less function file

> - an Excel custom function

Reviewer 2: I think each item in a list needs to be capitalized. The style guide doesn't explicitly state that, but every example is that way.

> Until now, if an add-in had more than one of these parts, then each part ran in its own separate JavaScript runtime, with its own global object and global variables.

Reviewer 2: "Until now" will get dated fast. Can we reference the requirement set instead?

> It's possible for add-ins with two or more parts to share a common JavaScript runtime. This shared runtime feature enables new preview APIs for hiding the task pane part of an add-in while the add-in remains running and for reopening the task pane later.

Reviewer 2: Suggestion for conciseness and clarity:
```suggestion
It's possible for add-ins with two or more parts to share a common JavaScript runtime. This shared runtime feature enables new preview APIs that hide and reopen the task pane while the add-in runs.
```

> Microsoft is now making it possible for add-ins with two or more parts to share a common JavaScript runtime. This new feature enables new preview APIs for hiding the task pane part of an add-in while the add-in remains running and for reopening the task pane later.

Reviewer: ```suggestion
It's possible for add-ins with two or more parts to share a common JavaScript runtime. This shared runtime feature enables new preview APIs for hiding the task pane part of an add-in while the add-in remains running and for reopening the task pane later.
```
I think I would mention that it's called shared runtime before the next ## section

> > [!IMPORTANT]

Reviewer: It might be worth adding an includes file specific to shared runtime (David might already have one?) so we don't have to update in multiple places

Author: Good idea. Done.

## Configure an add-in to use a shared runtime

To configure the add-in to use a shared runtime, see [Configure your Office Add-in to use a shared runtime](configure-your-add-in-to-use-a-shared-runtime.md).

## Show and hide the task pane

> The new APIs are in the `Office.addin` namespace. To show the task pane, your code calls `Office.addin.showAsTaskpane()`. Office will display in a task pane the page that you assigned to the resource ID (`resid`) for the task pane. This is the `resid` that you assigned to the `<SourceLocation>` of the `<Action xsi:type="ShowTaskpane">` in the manifest. (See [Configure your Office Add-in to use a shared runtime](configure-your-add-in-to-use-a-shared-runtime.md).)

Reviewer 2: `Office.addin` isn't technically a namespace, right? It's a read-only property, like `context`.

> The new APIs are in the `Office.addin` namespace. To show the task pane, your code calls `Office.addin.showAsTaskpane()`. Office will display in a task pane the page that you assigned to the resource ID (`resid`) for the task pane. This is the `resid` that you assigned to the `<SourceLocation>` of the `<Action xsi:type="ShowTaskpane">` in the manifest. (See [Configure your Office Add-in to use a shared runtime](configure-your-add-in-to-use-a-shared-runtime.md).)

Reviewer 2: I'm not sure you need to link the **Configure** article again. It's reference in the previous paragraph.

> This is an asynchronous method, so it your code should await it when the code that follows it should not run until it completes, either with the `await` keyword or with a `then()` method, depending on which JavaScript syntax you are using. The following assumes that there is an Excel worksheet named **CurrentQuarterSales**. The add-in should make the task pane visible whenever this worksheet is activated. The method `onCurrentQuarter` is a handler for the [Office.Worksheet.onActivated](/javascript/api/excel/excel.worksheet?view=excel-js-preview#onactivated) event which has been registered for the worksheet.

Reviewer 2: Typo? Also, small grammar suggestion.
```suggestion
This is an asynchronous method, so your code should await it when the subsequent code should not run until it completes. Wait for this completion with either the `await` keyword or a `then()` method, depending on which JavaScript syntax you are using. The following assumes that there is an Excel worksheet named **CurrentQuarterSales**. The add-in should make the task pane visible whenever this worksheet is activated. The method `onCurrentQuarter` is a handler for the [Office.Worksheet.onActivated](/javascript/api/excel/excel.worksheet?view=excel-js-preview#onactivated) event which has been registered for the worksheet.
```

> This is an asynchronous method, so code that should not run until it completes should be awaited, either with the `await` keyword or with a `then()` method, depending on which JavaScript syntax you are using. The following assumes that there is an Excel worksheet named **CurrentQuarterSales**. The task pane should be made visible whenever this worksheet is activated. The method `onCurrentQuarter` is a handler for the [Office.Worksheet.onActivated](/javascript/api/excel/excel.worksheet?view=excel-js-preview#onactivated) event which has been registered for the worksheet.

Reviewer: ```suggestion
This is an asynchronous method, so code that shouldn't run until it completes should be awaited, either with the `await` keyword or with a `then()` method, depending on which JavaScript syntax you are using. The following sample assumes that there is an Excel worksheet named **CurrentQuarterSales**. The task pane should be made visible whenever this worksheet is activated. The method `onCurrentQuarter` is a handler for the [Office.Worksheet.onActivated](/javascript/api/excel/excel.worksheet?view=excel-js-preview#onactivated) event which has been registered for the worksheet.
```
"should be awaited" & "should be made visible" are passive, consider rewriting to active voice

Author: done

function onCurrentQuarter() {
    Office.addin.showAsTaskpane()
    .then(function() {
        // Code that enables task pane UI elements for
        // working with the current quarter.
    });
}
```

To hide the task pane, your code calls `Office.addin.hide()`. The following example is a handler that is registered for the [Office.Worksheet.onDeactivated](/javascript/api/excel/excel.worksheet?view=excel-js-preview#ondeactivated) event.

```javascript
function onCurrentQuarterDeactivated() {
    Office.addin.hide();
}
```

### Preservation of state and event listeners

The `hide()` and `showAsTaskpane()` methods only change the *visibility* of the task pane. They do not unload or reload it (or reinitialize its state).

> Consider the following scenario: A task pane is designed with tabs. The **Home** tab is open when the add-in is first launched. Suppose a user opens the **Settings** tab and, later, code in the task pane calls `hide()` in response to some event. Still later code calls `showAsTaskpane()` in response to another event. The task pane will reappear, and the **Settings** tab is still elected.

Reviewer 2: Typo?
```suggestion
Consider the following scenario: A task pane is designed with tabs. The **Home** tab is open when the add-in is first launched. Suppose a user opens the **Settings** tab and, later, code in the task pane calls `hide()` in response to some event. Still later code calls `showAsTaskpane()` in response to another event. The task pane will reappear, and the **Settings** tab is still selected.
```

> Consider the following scenario: A task pane is designed with tabs. The **Home** tab is open when the add-in is first launched. Suppose a user opens the **Settings** tab and, later, code in the task pane calls `hide()` in response to some event. Still later code calls `showAsTaskpane()` in response to another event. The task pane will reappear, and the **Settings** tab is still elected.

Reviewer: Is there a way to illustrate the two scenarios through pictures? Or simplify the written descriptions? I get what they are saying but took me a couple times of moving my hands around to simulate the tabs to figure out what was going on. Not sure if this is specific to me as a physical/visual learner. I'm gonna also tag @AlexJerabek to see what he thinks on this one as I am dealing with some cognitive effects from a condition I have today and unsure if that is coming into play in my difficulty figuring this out quickly.

Reviewer: I almost don't know if you need the example given you say that it doesn't reload or change state in this first one

Author: Good catch. There was a graphic in the private preview Word doc, but I seem to have lost it when converting it to markdown. I've put it back in.

In addition, any event listeners that are registered in the task pane continue to run even when the task pane is hidden.

> Consider the following scenario: The task pane has a registered handler for the Excel `Worksheet.onActivated` and `Excel.Worksheet.onDeactivated` events for a sheet named **Sheet1**. The activated handler causes a green dot to appear in the task pane. The deactivated handler turns the dot red (which is its default state). Suppose then that code calls `hide()` when **Sheet1** is not activated and the dot is red. While the task pane is hidden, **Sheet1** is activated. Later code calls `showAsTaskpane()` in response to some event. When the task pane opens, the dot is green because the event listeners and handlers ran even though the task pane was hidden.

Reviewer 2: In this line, sometimes the `Excel` namespace is used and sometimes it's not. We should be consistent. I suggest omitting it.

### Handle visibility changed event

> Changing the visibility of the task pane is an event that your code can handle. To register a handler for the event, you do not use an "add handler" method as you would in most Office JavaScript contexts. Instead, there is a special function to which you pass your handler: `Office.addin.onVisibilityModeChanged`. The function returns another function that deregisters the handler. The following is an example. Note that the `visibilityMode` property of the `args` object that Office passes to the handler can have the values "Hidden" or "Taskpane".

Reviewer 2: I also find this example hard to parse. Is it possible to rewrite this as a bit of stubbed-out code?

Author: rewrote the whole section

> Changing the visibility of the task pane is an event that your code can handle. To register a handler for the event, you do not use an "add handler" method as you would in most Office JavaScript contexts. Instead, there is a special function to which you pass your handler: `Office.addin.onVisibilityModeChanged`. The function returns another function that deregisters the handler. The following is an example. Note that the `visibilityMode` property of the `args` object that Office passes to the handler can have the values "Hidden" or "Taskpane".

Reviewer: ```suggestion
You can change the visibility of a task pane through events. To register a handler for the event, don't use an "add handler" method. Instead, pass a special function to your handler: `Office.addin.onVisibilityModeChanged`. The function returns another function that deregisters the handler. 

In the following example, note that the `visibilityMode` property of the `args` object that Office passes to the handler can have the values "Hidden" or "Taskpane".
```

Author: Unfortunately, this suggestion changes the meaning. We don't want to say that the visibility changes in response to [unspecified] events. We want to say that a change in the visibility of the add-in (caused by the API, not an event) is itself an event that can be handled.

```javascript
var removeVisibilityModeHandler =
    await Office.addin.onVisibilityModeChanged(function(args) {
        console.log('Visibility is changed to ' + args.visibilityMode);
});
```

To deregister the handler:

```javascript

> await removeVisibilityModeHandler();

Reviewer 2: Are we officially okay with having `await` in the JS examples? If so, maybe we should consider cleaning up some older `then()` examples.

Author: rewrote whole section

![A screenshot of task pane that has four tabs labelled Home, Settings, Favorites, and Accounts.](../images/TaskpaneWithTabs.png)

In addition, any event listeners that are registered in the task pane continue to run even when the task pane is hidden.

Consider the following scenario: The task pane has a registered handler for the Excel `Worksheet.onActivated` and `Worksheet.onDeactivated` events for a sheet named **Sheet1**. The activated handler causes a green dot to appear in the task pane. The deactivated handler turns the dot red (which is its default state). Suppose then that code calls `hide()` when **Sheet1** is not activated and the dot is red. While the task pane is hidden, **Sheet1** is activated. Later code calls `showAsTaskpane()` in response to some event. When the task pane opens, the dot is green because the event listeners and handlers ran even though the task pane was hidden.

### Handle visibility changed event

> When your code changes the visibility of the task pane with `showAsTaskpane()` or `hide()`, Office triggers the VisibilityModeChanged event. It can be useful to handle this event. For example, suppose the task pane displays a list of all the sheets in a workbook. If a new worksheet is added while the task pane is hidden, making the task pane visible would not, in itself, add the new worksheet name to the list. But your code can respond to the VisibilityModeChanged event to reload the [Worksheet.name](/javascript/api/excel/excel.worksheet#name) property of all the worksheets in the [Workbook.worksheets](/javascript/api/excel/excel.workbook#worksheets) collection as shown in the example code below.

Reviewer 2: Missing code ticks around the event name.
```suggestion
When your code changes the visibility of the task pane with `showAsTaskpane()` or `hide()`, Office triggers the `VisibilityModeChanged` event. It can be useful to handle this event. For example, suppose the task pane displays a list of all the sheets in a workbook. If a new worksheet is added while the task pane is hidden, making the task pane visible would not, in itself, add the new worksheet name to the list. But your code can respond to the `VisibilityModeChanged` event to reload the [Worksheet.name](/javascript/api/excel/excel.worksheet#name) property of all the worksheets in the [Workbook.worksheets](/javascript/api/excel/excel.workbook#worksheets) collection as shown in the example code below.
```
